xapian-core/common/serialise-double.cc

```diff
@@ -38,175 +40,36 @@
 
 using namespace std;
 
-// The serialisation we use for doubles is inspired by a comp.lang.c post
-// by Jens Moeller:
-//
-// https://groups.google.com/group/comp.lang.c/browse_thread/thread/6558d4653f6dea8b/75a529ec03148c98
-//
-// The clever part is that the mantissa is encoded as a base-256 number which
-// means there's no rounding error provided both ends have FLT_RADIX as some
-// power of two.
-//
-// FLT_RADIX == 2 seems to be ubiquitous on modern UNIX platforms, while
-// some older platforms used FLT_RADIX == 16 (IBM machines for example).
-// FLT_RADIX == 10 seems to be very rare (the only instance Google finds
-// is for a cross-compiler to some TI calculators).
-
-#if FLT_RADIX == 2
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 7 + 7) / 8)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 8)
-# define MAX_MANTISSA (1 << (DBL_MAX_EXP & 7))
-#elif FLT_RADIX == 16
-# define MAX_MANTISSA_BYTES ((DBL_MANT_DIG + 1 + 1) / 2)
-# define MAX_EXP ((DBL_MAX_EXP + 1) / 2)
-# define MAX_MANTISSA (1 << ((DBL_MAX_EXP & 1) * 4))
-#else
-# error FLT_RADIX is a value not currently handled (not 2 or 16)
-// # define MAX_MANTISSA_BYTES (sizeof(double) + 1)
-#endif
-
-static int base256ify_double(double &v) {
-    int exp;
-    v = frexp(v, &exp);
-    // v is now in the range [0.5, 1.0)
-    --exp;
-#if FLT_RADIX == 2
-    v = scalbn(v, (exp & 7) + 1);
-#else
-    v = ldexp(v, (exp & 7) + 1);
+string serialise_double(double v)
+{
+    size_t double_size = sizeof(double);
+#ifdef WORDS_BIGENDIAN
+    uint64_t temp;
+    static_assert(sizeof(temp) == sizeof(v));
+    memcpy(&temp, &v, double_size);
```

Comment:
I think it's better to eliminate `double_size` - as a reviewer (and indeed as someone trying to follow the code) what I want to be able to see is that the right amount of data is being copied.  And while `double_size` suggests it holds the size of the double, I've seen enough misleadingly named variables in my time that I would want to check the definition - it's all-to-easy for someone to update the code but miss a place that should be changed.

In this version the size is clear from the line itself:

```
    memcpy(&temp, &v, sizeof(double));
```

I's also easier for the compiler to see the size - consider if the target type is too small:

```
#include <string.h>
#include <stdint.h>

uint32_t f(const char*p) {
    uint32_t r;
    memcpy(&r, p, sizeof(double));
    return r;
}

uint32_t g(const char*p) {
    const int double_size = sizeof(double);
    uint32_t r;
    memcpy(&r, p, double_size);
    return r;
}
```

Compiling that we get a warning when `sizeof(double)` is inlined but not if it comes through a variable (not even if I make it `const`):

```
$ gcc -Wall -W -c badcopy.c
badcopy.c: In function ‘f’:
badcopy.c:6:5: warning: ‘memcpy’ forming offset [5, 8] is out of the bounds [0, 4] of object ‘r’ with type ‘uint32_t’ {aka ‘unsigned int’} [-Warray-bounds]
     memcpy(&r, p, sizeof(double));
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
badcopy.c:5:14: note: ‘r’ declared here
     uint32_t r;
              ^
```

That's a potential stack smashing vulnerability.

Reply:
Fixed it. Thanks for the insight!